src/services/download/s3.rs

```diff
@@ -144,3 +144,224 @@ impl S3Downloader {
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::sources::{CommonSourceConfig, DirectoryLayoutType, SourceId};
+    use crate::test;
+    use crate::types::ObjectType;
+
+    use super::*;
+    use sha1::{Digest as _, Sha1};
+
+    const S3_BUCKET: &str = "symbolicator-test";
+
+    fn s3_source_key() -> Option<S3SourceKey> {
+        let access_key = std::env::var("AWS_ACCESS_KEY_ID").ok()?;
+        let secret_key = std::env::var("AWS_SECRET_ACCESS_KEY").ok()?;
+
+        if access_key.is_empty() || secret_key.is_empty() {
+            None
+        } else {
+            Some(S3SourceKey {
+                region: rusoto_core::Region::UsEast1,
+                access_key,
+                secret_key,
+            })
+        }
+    }
+
+    macro_rules! s3_source_key {
+        () => {
+            match s3_source_key() {
+                Some(key) => key,
+                None => {
+                    println!("Skipping due to missing SENTRY_SYMBOLICATOR_TEST_AWS_ACCESS_KEY_ID or SENTRY_SYMBOLICATOR_TEST_AWS_SECRET_ACCESS_KEY");
+                    return;
+                }
+            }
+        }
+    }
+
+    fn s3_client(source_key: S3SourceKey) -> rusoto_s3::S3Client {
+        rusoto_s3::S3Client::new_with(
+            rusoto_core::HttpClient::new().expect("create S3 HTTP client"),
+            rusoto_credential::StaticProvider::new_minimal(
+                source_key.access_key,
+                source_key.secret_key,
+            ),
+            source_key.region,
+        )
+    }
+
+    fn gcs_source(source_key: S3SourceKey) -> Arc<S3SourceConfig> {
+        Arc::new(S3SourceConfig {
+            id: SourceId::new("s3-test"),
+            bucket: S3_BUCKET.to_owned(),
+            prefix: String::new(),
+            source_key: Arc::new(source_key),
+            files: CommonSourceConfig::with_layout(DirectoryLayoutType::Unified),
+        })
+    }
+
+    #[test]
+    fn test_list_files() {
+        test::setup();
+
+        let source = gcs_source(s3_source_key!());
+        let downloader = S3Downloader::new();
+
+        let object_id = ObjectId {
+            code_id: Some("502fc0a51ec13e479998684fa139dca7".parse().unwrap()),
+            code_file: Some("Foo.app/Contents/Foo".to_owned()),
+            debug_id: Some("502fc0a5-1ec1-3e47-9998-684fa139dca7".parse().unwrap()),
+            debug_file: Some("Foo".to_owned()),
+            object_type: ObjectType::Macho,
+        };
+
+        let list = downloader.list_files(source, &[FileType::MachDebug], object_id);
+        assert_eq!(list.len(), 1);
+
+        assert_eq!(
+            list[0].location(),
+            SourceLocation::new("50/2fc0a51ec13e479998684fa139dca7/debuginfo")
+        );
+    }
+
+    /// Loads mock data into the S3 test bucket.
+    ///
+    /// This function will exit early if the bucket is already populated. Eventually, this should be
+    /// replaced with a one-off setup script.
+    async fn setup_bucket(source_key: S3SourceKey) {
```

Comment:
How well does this work when two tests are racing each other to set this up?  Is it reliable?  Maybe this should use a `std::sync::Once` to make it more explicitly safe?

Reply:
my question is rather: why do we need this?

Reply:
We tried this out, but it didn't give much of a speedup. On the contrary, if our setup code panics we will just receive a poison error instead of the verbose error message.

Reply:
@Swatinem we need this so that the tests are self-contained and externals can also test this easily with their own S3 accounts.